From: Utopia Maintenance Team <pkg-utopia-maintainers@lists.alioth.debian.org>
Date: Mon, 2 Oct 2017 10:30:41 +0300
Subject: _polkitunixsession_sessionid_from_display

---
 src/polkit/polkitunixsession-systemd.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/src/polkit/polkitunixsession-systemd.c b/src/polkit/polkitunixsession-systemd.c
index 8a8bf65..c34f36a 100644
--- a/src/polkit/polkitunixsession-systemd.c
+++ b/src/polkit/polkitunixsession-systemd.c
@@ -451,6 +451,7 @@ polkit_unix_session_initable_init (GInitable     *initable,
   PolkitUnixSession *session = POLKIT_UNIX_SESSION (initable);
   gboolean ret = FALSE;
   char *s;
+  uid_t uid;
 
   if (session->session_id != NULL)
     {
@@ -467,6 +468,19 @@ polkit_unix_session_initable_init (GInitable     *initable,
       goto out;
     }
 
+  /* Now do process -> uid -> graphical session (systemd version 213)*/
+  if (sd_pid_get_owner_uid (session->pid, &uid) < 0)
+    goto error;
+
+  if (sd_uid_get_display (uid, &s) >= 0)
+    {
+      session->session_id =  g_strdup (s);
+      free (s);
+      ret = TRUE;
+      goto out;
+    }
+
+error:
   g_set_error (error,
                POLKIT_ERROR,
                POLKIT_ERROR_FAILED,
